objax/util/check.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = ['assert_assigned_type_and_shape_match']
+
+import jax
+
+from objax.typing import JaxArray
+
+
+def split_shape_and_device(array):
+    if isinstance(array, jax.interpreters.pxla.ShardedDeviceArray):
+        return array.shape[0], array.shape[1:]
+    else:
+        return None, array.shape
+
+
+def assert_assigned_type_and_shape_match(existing_tensor, new_tensor):
```

Comment:
One concern is this function is super hard to understand. I'm not sure if we that's something we must accept or if it could be made simpler. If we don't think anything simpler at this point we could go with it and revisit in the future if we have new ideasa.

Reply:
I can add some comments to the code to explain why it's doing things. It's really easy to understand except for the jax tracing makes it all very complicated.

I could significantly simplify it by saying that if either type is a jax tracer just return true, but I tried to add some checks to this still.

Reply:
Well no need for now. Let's just move on with what we have right now.

For the failing test, maybe adjust the tolerance so that it passes.